plugins/pkg/file/fs.go

```diff
@@ -29,110 +29,88 @@ var (
 )
 
 type File struct {
-	lock sync.RWMutex
-
-	Name  string
-	mtime time.Time
-}
-
-func (f *File) Mtime() time.Time {
-	f.lock.RLock()
-	defer f.lock.RUnlock()
-	// the returned time.Time should be readonly
-	return f.mtime
+	Name    string
+	Watcher *fsnotify.Watcher
 }
 
-func (f *File) SetMtime(t time.Time) {
-	f.lock.Lock()
-	f.mtime = t
-	f.lock.Unlock()
+type StoreWatchedFiles struct {
+	WatchedFiles map[string]struct{}
+	lock         *sync.RWMutex
 }
 
-type fs struct {
-	cache *ttlcache.Cache[string, os.FileInfo]
-}
-
-func newFS(ttl time.Duration) *fs {
-	loader := ttlcache.LoaderFunc[string, os.FileInfo](
-		func(c *ttlcache.Cache[string, os.FileInfo], key string) *ttlcache.Item[string, os.FileInfo] {
-			info, err := os.Stat(key)
-			if err != nil {
-				logger.Error(err, "reload file info to cache", "file", key)
-				return nil
-			}
-			item := c.Set(key, info, ttlcache.DefaultTTL)
-			logger.Info("update file mtime", "file", key, "mtime", item.Value().ModTime())
-			return item
-		},
-	)
-	cache := ttlcache.New(
-		ttlcache.WithTTL[string, os.FileInfo](ttl),
-		ttlcache.WithLoader[string, os.FileInfo](loader),
-	)
-	go cache.Start()
-
-	return &fs{
-		cache: cache,
+func newStoreWatcherFiles() *StoreWatchedFiles {
+	return &StoreWatchedFiles{
+		WatchedFiles: make(map[string]struct{}),
+		lock:         &sync.RWMutex{},
 	}
 }
 
-var (
-	// TODO: rewrite it to use inotify
-	defaultFs = newFS(10 * time.Second)
-)
+var storeWatchedFiles = newStoreWatcherFiles()
 
-func IsChanged(files ...*File) bool {
-	for _, file := range files {
-		changed := defaultFs.isChanged(file)
-		if changed {
-			return true
+func WatchFiles(onChanged func(), file *File, otherFiles ...*File) (err error) {
+	files := append([]*File{file}, otherFiles...)
+	for _, f := range files {
+		if f == nil {
+			return errors.New("file pointer cannot be nil")
 		}
 	}
-	return false
-}
 
-func (f *fs) isChanged(file *File) bool {
-	item := f.cache.Get(file.Name)
-	if item == nil {
-		// As a protection, failed to fetch the real file means file not changed
-		return false
+	// Add files to watcher.
+	for _, f := range files {
+		go watchFiles(onChanged, f)
 	}
 
-	return file.Mtime().Before(item.Value().ModTime())
+	return
 }
 
-func (f *fs) Stat(path string) (*File, error) {
-	info, err := os.Stat(path)
-	if err != nil {
-		return nil, err
-	}
-	f.cache.Set(path, info, ttlcache.DefaultTTL)
+func watchFiles(onChanged func(), file *File) {
+	dir := filepath.Dir(file.Name)
+	defer func() {
+		storeWatchedFiles.lock.Lock()
```

Comment:
The global storeWatchedFiles makes things complex. And the current impl is incorrect if multiple config watch the same file.

We can improve it like this:
* create a watcher which wraps the fsnotify when initing the config
* add files into the watcher
* start the watcher (remember to stop the goroutine once the watcher is closed)